### PR TITLE
Implement double unlock checker

### PR DESCRIPTION
**Pull request by andersfischernielsen**
This will add two almost identical checkers to the code base along with test files for a confirmed positive from the Linux kernel code base. 

The only way the checkers differ is in whether they expect `may unlock --> !lock --> must unlock` or `must unlock --> !lock --> may unlock`.

### CODE REVIEW

**Comment by andersfischernielsen** [comment]
Only one on 15 true positives are detected by the checker as of writing. Here's the list of true positives found in the Linux kernel 

|  File                                       | Present in | Patched in |
|---------------------------------------------|------------|------------|
| drivers/block/drbd/drbd_main.c              | b0814361   | 8e9c5230   |
| fs/ubifs/orphan.c                           | 7542c6de   | 4dd75b33   | 
| drivers/gpu/drm/nouveau/nouveau_svm.c       | 5fbcf501   | de4ee728   |
| fs/btrfs/file.c                             | 78e03651   | f49aa1de   |
| drivers/staging/wilc1000/wilc_wlan.c        | ca641bae   | fea69916   |
| drivers/staging/kpc2000/kpc_dma/fileops.c   | d4c596eb   | c85aa326   |
| fs/nfs/client.c                             | a46126cc   | c260121a   |
| fs/btrfs/file.c                             | 2b90883c   | 8fca9550   |
| drivers/media/dvb-core/dvbdev.c             | ded71626   | 122d0e8d   |
| mm/memory_hotplug.c                         | 6376360e   | e3df4c6e   |
| sound/soc/codecs/pcm512x.c                  | fd270fca   | 28b698b7   |
| drivers/target/target_core_user.c           | 807cf197   | f0e89aae   |
| drivers/rpmsg/qcom_smd.c                    | fb416f69   | c3388a07   |
| drivers/scsi/aacraid/commsup.c              | 09624645   | d844752e   |
| drivers/staging/rtl8188eu/os_dep/usb_intf.c | 612e1c94   | 23bf4042   |
| block/blk-cgroup.c                          | e0223003   | bbb427e3   |

**Comment by andersfischernielsen** [comment]
I've now added GCC output for a true positive that this checker does *not* find.

**Comment by andersfischernielsen** [comment]
What the hell. I just debugged the entire solution and that example is correctly detected now. *Very* strange

**Comment by andersfischernielsen** [comment]
It seems that the `Q2_weak` check is where we lose the positives:
https://github.com/IagoAbal/eba/blob/dfb0c210776fdcc7d4583aa5667b671c593ba198/src/checkDUnlockFlow2.ml#L58

This seems strange to me since this simply uses the `mem` module of EBA. It doesn't seem to matter whether this is `mem_must` or `mem` (may). 

I also can't see how to make this more greedy to catch more positives (negative and positive)

File: src/type.ml [hunk 1]
```diff
@@ -1243,6 +1243,16 @@ and Effects : sig
 	= struct
 
 	type mem_kind = Alloc | Free | Read | Write | Uninit | Call | Lock | Unlock
+	
+	let mem_kind_pp m = match m with
```
Comment (IagoAbal) [inline src/type.ml:1247]: Did you use this function after all?

**Review by IagoAbal — CHANGES_REQUESTED** [review]
Hi, thanks for the pull request! I would like to avoid having three rather similar checkers. If there is any bug fix or improvement to make it would be nicer not having too much duplicate code. Could you create a common [functor](https://v1.realworldocaml.org/v1/en/html/functors.html) for the existing double-lock checker plus these new two? It could be a generic `DoubleOpChecker`. Then you can simply instantiate it three times.

**Comment by IagoAbal** [comment]
> Only one on 15 true positives are detected by the checker as of writing. Here's the list of true positives found in the Linux kernel

Remember to add `--all-lock-types` if you want EBA to search for other lock types than just spin-locks. Spin locks are simpler, EBA will produce more false positives for other lock types so that is disabled by default. Also EBA doesn't know all lock types in Linux, you may need to add some yourself.

**Pull request closed without merging**